Week 04 - Spring Core

sherenyang opened the pull request.

sherenyang committed:
Feat: Add assignment lecture 07

sherenyang committed:
Refactor: Update assignment lecture 07

sherenyang committed:
Refactor: Update assignment 3 lecture 07

ryuzakijebi commented:
The assignment is very well-executed, and the explanations are clear and easy to understand. Just a small note for Assignment 3.2: there is no need to specify server.port=8080 in application.properties, as 8080 is the default port for Spring Boot applications, unless you want to change the port.

sherenyang commented:
> The assignment is very well-executed, and the explanations are clear and easy to understand. Just a small note for Assignment 3.2: there is no need to specify server.port=8080 in application.properties, as 8080 is the default port for Spring Boot applications, unless you want to change the port.

Noted, thank you for the review!

sherenyang committed:
Feat: Add assignment lecture 08

helenhash commented on Week 04/Lecture 07/Assignment 2/Q1/assignment2/src/main/java/com/example/assignment2/config/AppConfig.java:
```diff
@@ -0,0 +1,9 @@
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@ComponentScan(basePackages = "com.example.assignment2")
```
Is there any case we can skip @componentscan?

sherenyang replied:
Yes, there is a case we can skip `@ComponentScan`. `@SpringBootApplication` already includes component scanning by default for the package where the main application class resides and all its sub-packages. We just need to ensure our main application class is in the base package.

Therefore, since we are using Spring Boot, we don’t need an additional `@ComponentScan` or `@Configuration` in `AppConfig` if our classes are in the right packages. Therefore, we remove `AppConfig` as it is not doing anything special.

I have revised my assignment and implemented this.

helenhash commented on Week 04/Lecture 08/Assignment 2/crud_employee/src/main/resources/application.properties:
```diff
@@ -0,0 +1,10 @@
+spring.datasource.password=
+spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver
+
+spring.jpa.hibernate.ddl-auto=update
```
What is the purpose of these 2 lines? We do not use jpa in this assignment, just make sure you understand.

sherenyang replied:
We do not use JPA in this assignment, so those two lines are unnecessary. I apologize for including them from my previous assignment. I have removed them in my revision. Thank you for reminding me!

helenhash commented on Week 04/Lecture 08/Assignment 3/crud_employee/src/main/java/com/example/crud_employee/config/DataSourceConfig.java:
```diff
@@ -0,0 +1,55 @@
+    @Bean(name = "dataSource1")
+    public DataSource dataSource1() {
+        return DataSourceBuilder.create()
+                .url("jdbc:mysql://localhost:3306/week4_employee")
```
All database info should placed in aplication.prop file and use @Value to get value, not hardcode here.

sherenyang replied:
I have moved all the database information in `application.properties` and use `@value` annotation in `DataSourceConfig` to get the value.

helenhash commented on Week 04/Lecture 08/Assignment 3/crud_employee/src/main/java/com/example/crud_employee/dao/EmployeeDaoImpl.java:
```diff
@@ -0,0 +1,81 @@
+        jdbcTemplate1.update(sql, employee.getId(), employee.getName(), employee.getDate_of_birth(), employee.getAddress(), employee.getDepartment());
+    }
+
+    @Transactional(transactionManager = "dataSource1TransactionManager")
```
It's good to @Transaction. 1 more request - can you research what is the case @Transactional does not work.

sherenyang replied:
I have updated my lecture 08 assignment 3 for `@Transactional` does not work case.

helenhash reviewed (changes requested):
Understand type of DI and know which one is best practice. Good explain about Circular DI, but should mention that try to avoid Cicular DI because it's really hard to debug and maintain.
I have some comments on PR below. Please check.

sherenyang committed:
Fix: Update assignment week 04 based on feedback

sherenyang commented:
Dear Ms. @helenhash,

Thank you for your feedback! I have updated the week 04 assignment based on your feedback. I also updated lecture 07 assignment 2 to mention that try to avoid Cicular DI because it's really hard to debug and maintain. Are all the revisions already alright?

Thank you for your guidance.

sherenyang committed:
Refactor: Update readme assignment 3 lecture 08
